docs/reference/performance.rst

```diff
@@ -0,0 +1,277 @@
+efficiency: serial to the host, parallel to the devices.
+
+For parallel workloads, when threads need to synchronize to share data, if they
+belong to the same block, they should use `__syncthreads()` (see:
+:ref:`synchronization functions`) within the same kernel invocation. If they 
+belong to different blocks, they must use global memory with two separate 
+kernel invocations. The latter should be minimized as it adds overhead.
+
+Device level
+-------------------------------------------------------------------------------
+Device-level optimization primarily involves maximizing parallel execution 
+across the multiprocessors of the device. This can be achieved by executing 
+multiple kernels concurrently on a device. The management of these kernels is
+facilitated by streams, which allow for the overlapping of computation and data
+transfers, enhancing performance. The aim is to keep all multiprocessors busy
+by executing enough kernels concurrently. However, launching too many kernels
+can lead to resource contention, so a balance must be found for optimal
+performance. This approach helps in achieving maximum utilization of the
+resources of the device.
+
+Multiprocessor level
+-------------------------------------------------------------------------------
+Multiprocessor-level optimization involves maximizing parallel execution within
+each multiprocessor on a device. Each multiprocessor can execute a number of 
+threads concurrently, and the total number of threads that can run in parallel
+is determined by the number of concurrent threads each multiprocessor can 
+handle.
+
+The key to multiprocessor-level optimization is to efficiently utilize the 
+various functional units within a multiprocessor. This can be achieved by 
+ensuring a sufficient number of resident warps, and at every instruction issue 
+time, a warp scheduler selects an instruction that is ready to execute. This 
+instruction can be another independent instruction of the same warp, exploiting
+:ref:`instruction optimization<instruction-level parallelism>`, or more 
+commonly an instruction of another warp, exploiting thread-level parallelism.
+
+In comparison, device-level optimization focuses on the device as a whole,
+aiming to keep all multiprocessors busy by executing enough kernels 
+concurrently. Both levels of optimization are crucial for achieving maximum 
+performance. They work together to ensure efficient utilization of the 
+resources of the GPU, from the individual multiprocessors to the device as a
+whole.
+
+.. _memory optimization:
+Memory optimization
+===============================================================================
+The first step in maximizing memory throughput is to minimize low-bandwidth 
+data transfers. This involves reducing data transfers between the host and the 
+device, as these have lower bandwidth than transfers between global memory and 
+the device.
```

Comment:
What is it "transfers between global memory and the device"? Global memory is a part of the device itself. Or you have in mind unified address space? Then it is not always true and should be rewritten more clearly anyway.

Reply:
Updated